ddl/table.go

```diff
@@ -1075,10 +1083,57 @@ func checkAndRenameTables(t *meta.Meta, job *model.Job, oldSchemaID, newSchemaID
 	err = updateLabelRules(job, tblInfo, oldRules, tableRuleID, partRuleIDs, oldRuleIDs, tblInfo.ID)
 	if err != nil {
 		job.State = model.JobStateCancelled
-		return ver, tblInfo, errors.Wrapf(err, "failed to update the label rule to PD")
+		return ver, errors.Wrapf(err, "failed to update the label rule to PD")
 	}
 
-	return ver, tblInfo, nil
+	return ver, nil
+}
+
+func adjustForeignKeyChildTableInfoAfterRenameTable(d *ddlCtx, t *meta.Meta, job *model.Job, tblInfo *model.TableInfo, oldSchemaName, oldTableName, newTableName model.CIStr, newSchemaID int64) ([]schemaIDAndTableInfo, error) {
+	if !variable.EnableForeignKey.Load() || newTableName.L == oldTableName.L {
+		return nil, nil
+	}
+	is, err := getAndCheckLatestInfoSchema(d, t)
+	if err != nil {
+		return nil, err
+	}
+	newDB, ok := is.SchemaByID(newSchemaID)
+	if !ok {
+		job.State = model.JobStateCancelled
+		return nil, infoschema.ErrDatabaseNotExists.GenWithStackByArgs(fmt.Sprintf("schema-ID: %v", newSchemaID))
+	}
+	infos := make(map[int64]schemaIDAndTableInfo)
+	fkh := newForeignKeyHelper(newDB.Name.L, newDB.ID, tblInfo)
+	referredFKs := is.GetTableReferredForeignKeys(oldSchemaName.L, oldTableName.L)
+	for _, referredFK := range referredFKs {
+		var childTableInfo schemaIDAndTableInfo
+		if referredFK.ChildSchema.L == oldSchemaName.L && referredFK.ChildTable.L == oldTableName.L {
+			childTableInfo = schemaIDAndTableInfo{newDB.ID, tblInfo}
+		} else {
+			childTableInfo, err = fkh.getTableFromStorage(is, t, referredFK.ChildSchema, referredFK.ChildTable)
+			if err != nil {
+				if infoschema.ErrTableNotExists.Equal(err) || infoschema.ErrDatabaseNotExists.Equal(err) {
+					continue
+				}
+				return nil, err
+			}
+		}
+		childFKInfo := model.FindFKInfoByName(childTableInfo.tblInfo.ForeignKeys, referredFK.ChildFKName.L)
+		if childFKInfo == nil {
+			continue
+		}
+		childFKInfo.RefSchema = newDB.Name
+		childFKInfo.RefTable = newTableName
+		infos[childTableInfo.tblInfo.ID] = childTableInfo
+	}
+	infoList := make([]schemaIDAndTableInfo, 0, len(fkh.loaded))
+	for _, info := range fkh.loaded {
+		if info.tblInfo.ID == tblInfo.ID {
```

Comment:
Will self table's reference be lost here?

Reply:
No, self table has already been updated, no need to update again.  the logic here is same with `adjustForeignKeyChildTableInfoAfterModifyColumn`

Reply:
<img width="1464" alt="image" src="https://user-images.githubusercontent.com/26020263/191699234-4dd0f0fd-2f57-4535-94cc-6a0f1f461a94.png">